### PR TITLE
8319196: ExecutableElement.getReceiverType doesn't return receiver types for methods loaded from bytecode

### PR DESCRIPTION
**Pull request by cushon**
Hello,

Please consider this fix for [JDK-8319196: ExecutableElement.getReceiverType doesn't return receiver types for methods loaded from bytecode](https://bugs.openjdk.org/browse/JDK-8319196).

As discussed in [the accompanying CSR](https://bugs.openjdk.org/browse/JDK-8319330), `ExecutableElement.getReceiverType()` and `ExecutableType.getReceiverType()` currently returns `NoType` unless the receiver parameter syntax was present in source, which also means that `getReceiverType()` doesn't work for methods loaded from class files.

This change updates the implementation to check if type annotations are present on the receiver type of a method read from bytecode, and fill in the receiver type of that method to host the type annotations.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8319330](https://bugs.openjdk.org/browse/JDK-8319330) to be approved

### Issues
 * [JDK-8319196](https://bugs.openjdk.org/browse/JDK-8319196): ExecutableElement.getReceiverType doesn't return receiver types for methods loaded from bytecode (**Bug** - P4)
 * [JDK-8319330](https://bugs.openjdk.org/browse/JDK-8319330): ExecutableElement.getReceiverType doesn't return receiver types for methods loaded from bytecode (**CSR**)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [32f0050a](https://git.openjdk.org/jdk/pull/16482/files/32f0050a70fe2b807090d5698dfd61339c503880)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16482/head:pull/16482` \
`$ git checkout pull/16482`

Update a local copy of the PR: \
`$ git checkout pull/16482` \
`$ git pull https://git.openjdk.org/jdk.git pull/16482/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16482`

View PR using the GUI difftool: \
`$ git pr show -t 16482`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16482.diff">https://git.openjdk.org/jdk/pull/16482.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16482#issuecomment-1791281930)

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java [hunk 1]
```diff
@@ -2276,6 +2277,21 @@ public Type getReceiverType() {
             return asType().getReceiverType();
         }
 
+        public Type implicitReceiverType() {
+            ClassSymbol enclosingClass = enclClass();
+            if (enclosingClass == null) {
+                return null;
+            }
+            Type enclosingType = enclClass().type;
```
Comment (liach) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java:2285]:
```suggestion
            Type enclosingType = enclosingClass.type;
```

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java [hunk 2]
```diff
@@ -2276,6 +2277,21 @@ public Type getReceiverType() {
...
+            if (isConstructor()) {
+                return enclosingType.getEnclosingType();
+            }
+            if (!isStatic()) {
```
Comment (liach) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java:2289]: Just curious, what happens in such a scenario:
```java
public class Outer {
  public static void work() {
    class Local {
      Local() {} // <- what's the implicitReceiverType here?
    }
  }
}
```
FYI javac currently is not marking such local classes in static methods as `static` in its `InnerClasses` attribute.

Comment (cushon) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java:2289]: Good question! I debugged that example, and `getEnclosingType()` returns `NoType`, which I think is what we want.

This is the logic that initializes the enclosing type and handles local classes:

https://github.com/openjdk/jdk/blob/c788160f8acea7b58b54ad857b601bb7ffb53f8e/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Enter.java#L508-L519

**Comment by cushon** [comment]
I have tentatively updated this to only fill in a receiver type during class reading if type annotations are present on the receiver, based on [discussion in the CSR](https://bugs.openjdk.org/browse/JDK-8319330?focusedId=14624193&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14624193).

File: test/langtools/tools/javac/processing/model/element/TestExecutableReceiverType.java [hunk 3]
```diff
@@ -45,8 +46,14 @@ public boolean process(Set<? extends TypeElement> annotations,
                            RoundEnvironment roundEnv) {
         if (!roundEnv.processingOver()) {
             int count = 0;
-            count += testType(elements.getTypeElement("MethodHost"));
-            count += testType(elements.getTypeElement("MethodHost.Nested"));
+            for (ExecutableElement e : ElementFilter.methodsIn(
+                  roundEnv.getElementsAnnotatedWith(ReceiverTypeKind.class))) {
+              count += testExecutable(e);
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/processing/model/element/TestExecutableReceiverType.java:51]: nit: indentation should be 4 spaces here, same issue below

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java [hunk 4]
```diff
@@ -81,6 +81,7 @@
 import static com.sun.tools.javac.jvm.ByteCodes.ishll;
 import static com.sun.tools.javac.jvm.ByteCodes.lushrl;
 import static com.sun.tools.javac.jvm.ByteCodes.lxor;
+import static com.sun.tools.javac.jvm.ByteCodes.ret;
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java:84]: this import doesn't seem to be used now

**Review by vicente-romero-oracle — APPROVED** [review]
looks good to me, just minor comments

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java (hunk 4 above)
Comment (cushon) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java]: Done

File: test/langtools/tools/javac/processing/model/element/TestExecutableReceiverType.java (hunk 3 above)
Comment (cushon) [reply test/langtools/tools/javac/processing/model/element/TestExecutableReceiverType.java]: Done

**Comment by cushon** [comment]
> looks good to me, just minor comments

I addressed the comments, thanks for the review!

I also updated the CSR [JDK-8319330](https://bugs.openjdk.org/browse/JDK-8319330) and moved it back to 'proposed'.

**Pull request closed without merging**